File: backend/app/schemas/nlp.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class AnalisisTextoRequest(BaseModel):
    texto: str = Field(
        min_length=1
    )


class PalabraFrecuente(BaseModel):
    palabra: str
    frecuencia: int


class AnalisisTextoResponse(BaseModel):
    cantidad_palabras: int
    tokens: list[str]
    palabras_frecuentes: list[PalabraFrecuente]


class ClasificacionRequest(BaseModel):
    texto: str = Field(
        min_length=1
    )


class ClasificacionResponse(BaseModel):
    categoria: str
    confianza: float


class AnalisisNLPResponse(BaseModel):
    id: int
    comentario_id: int

    idioma: str
    cantidad_palabras: int

    palabras_limpias: list[str] | None = None
    palabras_frecuentes: list[PalabraFrecuente] | None = None

    categoria_detectada: str | None = None
    confianza: float | None = None

    fecha_analisis: datetime | None = None

    model_config = ConfigDict(
        from_attributes=True
    )

File: frontend/src/services/nlp.ts
import { apiFetch } from "./api";


export interface PalabraFrecuente {
  palabra: string;
  frecuencia: number;
}


export interface AnalisisTexto {
  cantidad_palabras: number;
  tokens: string[];
  palabras_frecuentes: PalabraFrecuente[];
}


export interface Clasificacion {
  categoria: string;
  confianza: number;
}


export interface AnalisisNLP {
  id: number;
  comentario_id: number;
  idioma: string;
  cantidad_palabras: number;

  palabras_limpias: string[] | null;

  palabras_frecuentes:
    | PalabraFrecuente[]
    | null;

  categoria_detectada:
    | string
    | null;

  confianza:
    | number
    | null;

  fecha_analisis:
    | string
    | null;
}


export function analizarTexto(
  texto: string
) {
  return apiFetch<AnalisisTexto>(
    "/api/nltk/analizar",
    {
      method: "POST",
      body: JSON.stringify({
        texto,
      }),
    }
  );
}


export function clasificarTexto(
  texto: string
) {
  return apiFetch<Clasificacion>(
    "/api/nltk/clasificar",
    {
      method: "POST",
      body: JSON.stringify({
        texto,
      }),
    }
  );
}


export function analizarComentarioGuardado(
  comentarioId: number
) {
  return apiFetch<AnalisisNLP>(
    `/api/nltk/comentarios/${comentarioId}/analizar`,
    {
      method: "POST",
    }
  );
}


export function obtenerAnalisisNLP() {
  return apiFetch<AnalisisNLP[]>(
    "/api/nltk/analisis"
  );
}